simplecnf.go

```diff
@@ -0,0 +1,160 @@
+		m[k] = v
+	}
+}
+
+func (f simpleSatCnf) UnitPropagate() (CnfFormula, map[Variable]bool, []Variable) {
+	assignments := variableAssignment(make(map[Variable]bool))
+	formula := f
+	for {
+		var a map[Variable]bool
+		var conflicts []Variable
+		formula, a, conflicts = formula.unitPropagate1()
+		if len(a) == 0 || len(conflicts) != 0 {
+			return formula, map[Variable]bool(assignments), conflicts
+		} else {
+			assignments.putAll(a)
+		}
+	}
+}
+
+// unitPropagate1 assigns singleton clauses to the values that they must be.
+// Unlike UnitPropagate, this method does not recurse and assign unit variables again
+func (f simpleSatCnf) unitPropagate1() (simpleSatCnf, map[Variable]bool, []Variable) {
+	assignments, conflicts := f.getUnitPropagateAssignments()
+	return f.assignVariables(assignments), assignments, conflicts
+}
+
+// getUnitPropagateAssignments figures out what variables have been pinned down
+// into unit clauses and need some assignment to work.
+func (f simpleSatCnf) getUnitPropagateAssignments() (assignments map[Variable]bool, conflicts []Variable) {
+	assignments = make(map[Variable]bool)
+	for _, c := range f.clauses {
+		if len(c) == 1 {
+			assignedVar := c[0]
+			variable := assignedVar.Variable()
+			prevAssignment, ok := assignments[variable]
+			if ok  {
+				if prevAssignment == assignedVar.Value() && !containsVariable(conflicts, variable) {
+					conflicts = append(conflicts, variable)
+				}
+			} else {
+				assignments[variable] = assignedVar.Value()
+			}
+		}
+	}
+	// TODO: sort conflicts
+	for _, con := range conflicts {
+		delete(assignments, con)
+	}
+	return assignments, conflicts
+}
+
+// containsVariable returns true if the second argument is in the first slice
+func containsVariable(variables []Variable, v Variable) bool {
+	for _, s := range variables {
+		if s == v {
+			return true
+		}
+	}
+	return false
+}
+
+func (f simpleSatCnf) PureVariableAssign() (CnfFormula, map[Variable]bool) {
+	// TODO: implement.  Current implementation will produce correct result, but slower
+	return f, nil
+}
+
+// Selects a variable to guess on.
+// Precondition: the clause is not unsatisfyible or emtpy (satisfied)
+func (f simpleSatCnf) ChooseVariable() Variable {
+	// The first var of the first clause that has the fewest number of variables
+	minLength := math.MaxInt64
+	var v Variable
+	for _, c := range f.clauses {
+		if 0 < len(c) && len(c) < minLength {
+			minLength = len(c)
+			v = c[0].Variable()
+		}
+	}
+	return v
+}
+
+func (f simpleSatCnf) AssignVariables(assignments map[Variable]bool) CnfFormula {
+	return f.assignVariables(assignments)
+}
+
+func (f simpleSatCnf) assignVariables(assignments map[Variable]bool) simpleSatCnf {
+	if len(assignments) == 0 {
+		return f
+	}
+	newClauses := make([]orClause, 0, len(f.clauses))
+clauseLoop:
+	for _, c := range f.clauses {
+		var varsToRemoveFromClause []Variable
+		for _, negVar := range c {
+			variable := negVar.Variable()
+			if ass, ok := assignments[variable]; ok {
```

Comment:
ass is not a great variable name either.

Reply:
should there be an else clause here, for the case where assignments[variable] is not ok?

Reply:
i believe that we don't need an else clause, but this is not intuitive. perhaps a thorough comment.